packages/playwright-test/src/reporters/blob.ts

```diff
@@ -0,0 +1,184 @@
+    if (!this._outputFile)
+      return;
+
+    const reportString = JSON.stringify(this._messages, undefined, 2);
+    fs.mkdirSync(path.dirname(this._outputFile), { recursive: true });
+    await zipReport(this._outputFile, reportString);
+  }
+
+  private _resolveOutputDir(config: FullConfigInternal): string {
+    const { outputDir } = this._options;
+    if (outputDir)
+      return path.resolve(config._internal.configDir, outputDir);
+    return defaultReportFolder(config._internal.configDir);
+  }
+}
+
+export async function createMergedReport(config: FullConfigInternal, dir: string, reporterName?: string) {
+  const shardFiles = await sortedShardFiles(dir);
+  console.log(`Merging ${shardFiles}`);
+  const events = await mergeEvents(dir, shardFiles);
+  console.log(`Writing ${reporterName} report to ${dir}`);
+
+  // TODO: use reporterName.
+  const reporter = new HtmlReporter({ _internalResolvedOutputFolder: dir });
+  const receiver = new TeleReporterReceiver(path.sep, reporter);
+  // TODO: handle onError.
+  for (const event of events)
+    await receiver.dispatch(event);
+  console.log(`Done.`);
+}
+
+async function mergeEvents(dir: string, shardFiles: string[]) {
+  const events: JsonEvent[] = [];
+  const beginEvents: JsonEvent[] = [];
+  const endEvents: JsonEvent[] = [];
+  for (const file of shardFiles) {
+    const zipFile = new ZipFile(path.join(dir, file));
+    const entryNames = await zipFile.entries();
+    const reportEntryName = entryNames.find(e => e.endsWith('.json'));
+    if (!reportEntryName)
+      throw new Error(`Zip file ${file} does not contain a .json file`);
+    const reportJson = await zipFile.read(reportEntryName);
+    const parsedEvents = JSON.parse(reportJson.toString()) as JsonEvent[];
+    for (const event of parsedEvents) {
+      // TODO: show remaining events?
+      if (event.method === 'onError')
+        throw new Error('Error in shard: ' + file);
+      if (event.method === 'onBegin')
+        beginEvents.push(event);
+      else if (event.method === 'onEnd')
+        endEvents.push(event);
+      else
+        events.push(event);
+    }
+
+  }
+  return [mergeBeginEvents(beginEvents), ...events, mergeEndEvents(endEvents)];
+}
+
+function mergeBeginEvents(beginEvents: JsonEvent[]): JsonEvent {
+  if (!beginEvents.length)
+    throw new Error('No begin events found');
+  const projects: JsonProject[] = [];
+  for (const event of beginEvents) {
+    const shardProjects: JsonProject[] = event.params.projects;
+    for (const shardProject of shardProjects) {
+      const mergedProject = projects.find(p => p.id === shardProject.id);
+      if (!mergedProject)
+        projects.push(shardProject);
+      else
+        mergeJsonSuites(shardProject.suites, mergedProject);
+    }
+  }
+  return {
+    method: 'onBegin',
+    params: {
+      config: beginEvents[0].params.config,
+      projects,
+    }
+  };
+}
+
+function mergeJsonSuites(jsonSuites: JsonSuite[], parent: JsonSuite | JsonProject) {
```

Comment:
There is similar code in the telereceiver, what is the difference.

Reply:
This one operates on Json* structures while that works with Tele* and the logic is different in many ways.